lib/eigensolve_quda.cpp

```diff
@@ -0,0 +1,948 @@
+        }
+      }
+    } else {
+      while (i < nKr) {
+        if ((i == 0) || (alpha[i - 1] <= alpha[i]))
+          i++;
+        else {
+          double tmp = alpha[i];
+          alpha[i] = alpha[i - 1];
+          alpha[--i] = tmp;
+          *r[0] = *kSpace[i];
+          *kSpace[i] = *kSpace[i - 1];
+          *kSpace[i - 1] = *r[0];
+        }
+      }
+    }
+  }
+
+  void TRLM::eigensolveFromArrowMat(int num_locked, int arrow_pos)
+  {
+
+    int dim = nKr - num_locked;
+
+    time_ = -clock();
+    // Eigen objects
+    MatrixXd A = MatrixXd::Zero(dim, dim);
+    ritz_mat.resize(dim * dim);
+    for (int i = 0; i < dim * dim; i++) ritz_mat[i] = 0.0;
+
+    // Invert the spectrum due to chebyshev
+    if (reverse) {
+      for (int i = num_locked; i < nKr - 1; i++) {
+        // printfQuda("Alpha[%d] = %e, beta[%d] = %e\n", i, alpha[i], i, beta[i]);
+        alpha[i] *= -1.0;
+        beta[i] *= -1.0;
+      }
+      alpha[nKr - 1] *= -1.0;
+    }
+
+    // Construct arrow mat A_{dim,dim}
+    for (int i = 0; i < dim; i++) {
+
+      // alpha populates the diagonal
+      A(i, i) = alpha[i + num_locked];
+    }
+
+    for (int i = 0; i < arrow_pos - 1; i++) {
+
+      // beta populates the arrow
+      A(i, arrow_pos - 1) = beta[i + num_locked];
+      A(arrow_pos - 1, i) = beta[i + num_locked];
+    }
+
+    for (int i = arrow_pos - 1; i < dim - 1; i++) {
+
+      // beta populates the sub-diagonal
+      A(i, i + 1) = beta[i + num_locked];
+      A(i + 1, i) = beta[i + num_locked];
+    }
+
+    // Eigensolve the arrow matrix
+    SelfAdjointEigenSolver<MatrixXd> eigensolver;
+    eigensolver.compute(A);
+
+    // repopulate ritz matrix
+    for (int i = 0; i < dim; i++)
+      for (int j = 0; j < dim; j++) ritz_mat[dim * i + j] = eigensolver.eigenvectors().col(i)[j];
+
+    for (int i = 0; i < dim; i++) {
+      residua[i + num_locked] = fabs(beta[nKr - 1] * eigensolver.eigenvectors().col(i)[dim - 1]);
+      // Update the alpha array
+      alpha[i + num_locked] = eigensolver.eigenvalues()[i];
+    }
+
+    // Put spectrum back in order
+    if (reverse) {
+      for (int i = num_locked; i < nKr; i++) { alpha[i] *= -1.0; }
+    }
+
+    time_ += clock();
+    time_e += time_;
+  }
+
+  void TRLM::computeKeptRitz(std::vector<ColorSpinorField *> &kSpace)
+  {
+
+    int offset = nKr + 1;
+    int dim = nKr - num_locked;
+
+    if ((int)kSpace.size() < offset + iter_keep) {
+      for (int i = kSpace.size(); i < offset + iter_keep; i++) {
+        if (getVerbosity() >= QUDA_DEBUG_VERBOSE) printfQuda("Adding %d vector to kSpace\n", i);
+        kSpace.push_back(ColorSpinorField::Create(csParam));
+      }
+    }
+
+    for (int i = 0; i < iter_keep; i++) {
+      int k = offset + i;
+      *r[0] = *kSpace[num_locked];
+      blas::ax(ritz_mat[dim * i], *r[0]);
+      *kSpace[k] = *r[0];
+      for (int j = 1; j < dim; j++) { blas::axpy(ritz_mat[i * dim + j], *kSpace[num_locked + j], *kSpace[k]); }
```

Comment:
Not that I want to add more multi-blas routines, but for performance reasons would it ultimately be a net win to replace this with a batch `caxpy` as opposed to multiple `axpy`s? (Bandwidth savings compensating for unnecessary arithmetic costs)

Reply:
At least be a net win*

Reply:
In the `axpy`, can the indices `k` and `num_locked + j` be equal?  Just wondering hos easy this is multi-blas-ify

Reply:
This can multi-blas-ified. `num_locked + j` is never greater than `dim`, which is never greater than `nKr`. `k` is always equal to `nKr + 1`, so the vectors never alias.